src/flag_gems/ops/dropout.py

```diff
@@ -4,83 +4,104 @@
-    _seed, _offset = philox_cuda_seed_offset(0)
-    _rand[_grid](_seed, _offset)
-except Exception:
-    tl_rand_dtype = tl.int32
-
-del _grid
-del _seed
-del _offset
-
-
-@libentry()
-@triton.autotune(
-    configs=[
-        triton.Config({"N_BLOCK_SIZE": 256}, num_warps=2, num_stages=4),
-        triton.Config({"N_BLOCK_SIZE": 256}, num_warps=2, num_stages=5),
-        triton.Config({"N_BLOCK_SIZE": 512}, num_warps=2, num_stages=4),
-        triton.Config({"N_BLOCK_SIZE": 512}, num_warps=2, num_stages=5),
-        triton.Config({"N_BLOCK_SIZE": 1024}, num_warps=4, num_stages=4),
-        triton.Config({"N_BLOCK_SIZE": 1024}, num_warps=4, num_stages=5),
-        triton.Config({"N_BLOCK_SIZE": 2048}, num_warps=4, num_stages=4),
-        triton.Config({"N_BLOCK_SIZE": 2048}, num_warps=4, num_stages=5),
-    ],
-    key=[
-        "N",
-    ],
+    def uint_to_uniform_float(x):
+        """
+        Numerically stable function to convert a random uint into a random float uniformly sampled in [0, 1).
+        """
+        # TODO: fix frontend issues and cleanup
+        # conditions can be simplified
+        # scale is ((2**23 - 1) / 2**23) * 2**(N_BITS - 1)
+        if tl.constexpr(x.dtype == tl.uint32) or tl.constexpr(x.dtype == tl.int32):
+            # maximum value such that `MAX_INT * scale < 1.0` (with float rounding)
+            x = x.to(tl.int32, bitcast=True)
+            scale = 4.6566127342e-10
+        else:
+            tl.static_assert(
+                tl.constexpr(x.dtype == tl.uint64) or tl.constexpr(x.dtype == tl.int64)
+            )
+            x = x.to(tl.int64, bitcast=True)
+            scale = 1.0842020432385337e-19
+        x = tl.where(x < 0, -x - 1, x)
+        return x * scale
+
+
+UNROLL = 4
+
+
+@triton.heuristics(
+    values={
+        # "BLOCK": lambda args: 512 if args["N"] <= 512 else 1024 if args["N"] <= 1024 else 2048,
+        "BLOCK": lambda args: 512 if args["N"] <= 512 else 1024,
+        "num_warps": lambda args: 4
+        if args["N"] <= 512
+        else 8
+        if args["N"] <= 1024
+        else 16,
+    }
 )
 @triton.jit
-def dropout_forward_kernel(
-    X,
-    Y,
-    N,
-    p,
-    philox_seed,
-    philox_offset,
-    N_BLOCK_SIZE: tl.constexpr,
-):
+def dropout_forward_kernel(X, Y, N, p, philox_seed, philox_offset, BLOCK: tl.constexpr):
+    UNROLL = 4
     philox_seed = philox_seed.to(tl.int64)
-    philox_offset = philox_offset.to(tl_rand_dtype)
-    pid = tl.program_id(0) * N_BLOCK_SIZE
-    offset = pid + tl.arange(0, N_BLOCK_SIZE)
-    mask = offset < N
-    X_ptr = X + offset
-    Y_ptr = Y + offset
-    inp = tl.load(X_ptr, mask=mask, other=0.0)
-    philox_offset = philox_offset + offset
-    pmask = tl.rand(philox_seed, philox_offset, n_rounds=6) > p
+    philox_offset = philox_offset.to(tl.int64)
+    c0 = (philox_offset & 0xFFFFFFFF).to(tl.uint32)
+    c1 = ((philox_offset >> 32) & 0xFFFFFFFF).to(tl.uint32)
+    i4 = tl.program_id(0) * BLOCK + tl.arange(0, BLOCK)
```

Comment:
So `i4` is the inner-CTA offset of the counter?

Reply:
May be we can use 2 64-bit counter, one for each invocation of the kernel(increment by 1 for each invocation), and the other for the in-kernel task id. Then we can make full use of the 128 bit counter.

Reply:
i4 represents offsets that increments one by every UNROLL(=4) elements.